File: src/MyWebLog.DataCS/Category.cs
namespace MyWebLog.Data;

/// <summary>
/// A category under which a post may be identfied
/// </summary>
public class Category
{
    /// <summary>
    /// The ID of the category
    /// </summary>
    public string Id { get; set; } = "";

    /// <summary>
    /// The displayed name
    /// </summary>
    public string Name { get; set; } = "";

    /// <summary>
    /// The slug (used in category URLs)
    /// </summary>
    public string Slug { get; set; } = "";

    /// <summary>
    /// A longer description of the category
    /// </summary>
    public string? Description { get; set; } = null;

    /// <summary>
    /// The parent ID of this category (if a subcategory)
    /// </summary>
    public string? ParentId { get; set; } = null;

    /// <summary>
    /// The parent of this category (if a subcategory)
    /// </summary>
    public Category? Parent { get; set; } = default;

    /// <summary>
    /// The posts assigned to this category
    /// </summary>
    public ICollection<Post> Posts { get; set; } = default!;
}

File: src/MyWebLog/Handlers/Category.fs
/// Handlers to manipulate categories
module MyWebLog.Handlers.Category

open DotLiquid
open Giraffe
open MyWebLog

// GET /categories
let all : HttpHandler = requireUser >=> fun next ctx -> task {
    return!
        Hash.FromAnonymousObject {|
            categories = CategoryCache.get ctx
            page_title = "Categories"
            csrf       = csrfToken ctx
        |}
        |> viewForTheme "admin" "category-list" next ctx
}

open MyWebLog.ViewModels

// GET /category/{id}/edit
let edit catId : HttpHandler = requireUser >=> fun next ctx -> task {
    let  webLogId = webLogId ctx
    let  conn     = conn     ctx
    let! result   = task {
        match catId with
        | "new" -> return Some ("Add a New Category", { Category.empty with id = CategoryId "new" })
        | _ ->
            match! Data.Category.findById (CategoryId catId) webLogId conn with
            | Some cat -> return Some ("Edit Category", cat)
            | None -> return None
    }
    match result with
    | Some (title, cat) ->
        return!
            Hash.FromAnonymousObject {|
                csrf       = csrfToken ctx
                model      = EditCategoryModel.fromCategory cat
                page_title = title
                categories = CategoryCache.get ctx
            |}
            |> viewForTheme "admin" "category-edit" next ctx
    | None -> return! Error.notFound next ctx
}

// POST /category/save
let save : HttpHandler = requireUser >=> validateCsrf >=> fun next ctx -> task {
    let! model    = ctx.BindFormAsync<EditCategoryModel> ()
    let  webLogId = webLogId ctx
    let  conn     = conn     ctx
    let! category = task {
        match model.categoryId with
        | "new" -> return Some { Category.empty with id = CategoryId.create (); webLogId = webLogId }
        | catId -> return! Data.Category.findById (CategoryId catId) webLogId conn
    }
    match category with
    | Some cat ->
        let cat =
            { cat with
                name        = model.name
                slug        = model.slug
                description = if model.description = "" then None else Some model.description
                parentId    = if model.parentId    = "" then None else Some (CategoryId model.parentId)
            }
        do! (match model.categoryId with "new" -> Data.Category.add | _ -> Data.Category.update) cat conn
        do! CategoryCache.update ctx
        do! addMessage ctx { UserMessage.success with message = "Category saved successfully" }
        return! redirectToGet $"/category/{CategoryId.toString cat.id}/edit" next ctx
    | None -> return! Error.notFound next ctx
}

// POST /category/{id}/delete
let delete catId : HttpHandler = requireUser >=> validateCsrf >=> fun next ctx -> task {
    let webLogId = webLogId ctx
    let conn     = conn     ctx
    match! Data.Category.delete (CategoryId catId) webLogId conn with
    | true ->
        do! CategoryCache.update ctx
        do! addMessage ctx { UserMessage.success with message = "Category deleted successfully" }
    | false -> do! addMessage ctx { UserMessage.error with message = "Category not found; cannot delete" }
    return! redirectToGet "/categories" next ctx
}
